instrumentation/reactor-netty-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactornetty/ReactorNettyInstrumentationModule.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.reactornetty;
+
+import static java.util.Collections.singletonList;
+import static java.util.Collections.singletonMap;
+import static net.bytebuddy.matcher.ElementMatchers.isStatic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.javaagent.instrumentation.api.CallDepthThreadLocalMap;
+import io.opentelemetry.javaagent.instrumentation.netty.v4_1.AttributeKeys;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import reactor.core.publisher.Mono;
+import reactor.netty.Connection;
+import reactor.netty.http.client.HttpClient;
+import reactor.netty.http.client.HttpClientRequest;
+
+/**
+ * This instrumentation solves the problem of the correct context propagation through the roller
+ * coaster of Project Reactor and Netty thread hopping. It uses two public hooks of {@link
+ * HttpClient}: {@link HttpClient#mapConnect(BiFunction)} and {@link
+ * HttpClient#doOnRequest(BiConsumer)} two short-cut context propagation from the caller to Reactor
+ * to Netty.
+ */
+@AutoService(InstrumentationModule.class)
+public class ReactorNettyInstrumentationModule extends InstrumentationModule {
+
+  public ReactorNettyInstrumentationModule() {
+    super("reactor-netty", "reactor-netty-0.9");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return singletonList(new HttpClientInstrumentation());
+  }
+
+  public static class HttpClientInstrumentation implements TypeInstrumentation {
+    @Override
+    public ElementMatcher<TypeDescription> typeMatcher() {
+      return named("reactor.netty.http.client.HttpClient");
```

Comment:
since both 0.9 and 1.0 apply to the same class, would be good to make sure that they don't conflict, e.g.
* add muzzle `assertInverse` if possible
* add a `testInstrumentation` dependency from 1.0 on 0.9 and vice versa (we do this in a couple of places, but should probably do it uniformly wherever we have instrumentation split across library versions)

Reply:
Also debug logging will be confusing, it'd be nice to have a `classLoaderMatcher` to differentiate rather than only use the muzzle check

Reply:
Good points, will do.

Reply:
Done